### DISCUSSION TITLE
Converting archived hubs to fully-configured hubVerse-formatted hubs

### DISCUSSION THREAD

**Discussion by lmullany** [hubverse-org/.github#7] [General]
We are going to try to convert one or more achived/no-long-active hubs to hubverse format. The first one that we will attempt to do this for is the 2015-2019 FluSight hub.

**Comment by bsweger** [comment]
@lmullany Sounds great! @nickreich and I were talking about how this effort ties in nicely to the ongoing infrastructure work for cloud-enabled hubs.

Because there are no active submissions to worry about, we could use the data you're creating to test the cloud sync and parquet conversions without the worry of disrupting a live hub.

Pushing this historic data to the cloud would also provide a tangible talking point for our conversations about cloud data access, optimization, etc

Very excited!

**Comment by lmullany** [comment]
@bsweger @nickreich:  the initial version of the hubverse-formatted flusight1 repo is here:

https://github.com/lmullany/flusight1_hub

This should be considered preliminary. There are a number of issues that we might want to sort through including:
1. Is the selection of the `origin_epiweek`, as the main round identifier, the right choice?
2. How should we handle in-valid data? For example, we have model submissions where the cdf max exceeds 1, or the pmf does not sum to 1
3. Some teams changed their name from year to year, and thus in `model-output` can have multiple folders, but if this is truly a name change only, perhaps a common name should be chosen and the multiple folders consolidated into on
4. There is nothing in model-metadata folder, because the original repo did not contain a metadata information file for each of the models.
5. Occaisionally teams submitted two .csv files for the same EWXX value. For example, we can have something like `EW50_Team_model_2019-12-25.csv` and `EW50_Team_model_2019-12-27.csv`. In these situations, I took the earliest file (i.e. the `EW50` file with the `2019-12-25`  date embedded in the file name

Maybe we can setup a time to discuss further

**Reply by nickreich** [reply]
Thanks for the progress on this! I don't have access to the repo. I think some of the issues you list might be more thorny than others. I might suggest making separate issues in the repo you created for each of the above issues and then we could have threaded conversations about each of them.

**Reply by lmullany** [reply]
Will do

**Reply by lmullany** [reply]
@nickreich I've added issues to the github for the above concerns

**Comment by bsweger** [comment]
Very exciting--thanks, @lmullany !

I'll await the more specific issues before chiming in, but in the meantime, I'm continuing to codify the function that converts model-output data to parquet (add adds columns to round_id, team, and model) so it will trigger automatically when a hub's data is synced to the cloud.

This dovetails nicely with the work you're doing. Once we have an official hubverse-compatible version of the flusight archive, we can add the cloud infrastructure pieces to make the data accessible via S3. 🚀

**Reply by lmullany** [reply]
Here, team and model are not well distinguished. While model metadata in a “typical” repo that is created ( and therefore validated) from the outset using hubverse tools require this distinction ( I think!), that may not necessarily be true with post hoc created repo like this one. 

So we might want to consider designing the functionality you mention above with some degree of flexibility with regards to the  structure of `model-output` data and the presence/absence of information in the `model-metadata` folder

**Reply by bsweger** [reply]
Interesting--thanks for the heads-up. 

The [transform function's unit of work is a single model-output file](https://github.com/Infectious-Disease-Modeling-Hubs/hubverse-infrastructure/blob/bsweger/add-model-output-transforms/hubverse-transforms/src/hubverse_transforms/model_output.py#L18), which allows us to use S3 event notifications + AWS Lambda to transform each model-output file after it arrives in the bucket. 

Your comment is a good note to parse the team and model name via directory structure and not filename (the S3 event sends the entire object key, so that's not a problem). Appreciate it!

https://github.com/Infectious-Disease-Modeling-Hubs/hubverse-transform/issues/2

**Comment by lmullany** [comment]
@bsweger the current version of the repo is structured like this:

![image](https://github.com/Infectious-Disease-Modeling-Hubs/.github/assets/12236192/a4bbd78e-050a-49a5-a6ea-33a040ecec01)

Note that `model-output` folder has a subfolder for each model (the name of that folder is sometimes in team-model format (for example `"NEU-GLEAM"`, but in many instances that type of information was not provided, so we have folder names that really just describe the model itself

Also note that within the sub-folder, we have a list of parquet files, each one labeled as "YYYY-EW", where EW is the epiweek. This 7 character string is also the `round_id`. (That column is included in the parquet file as well as a constant). 

Does the constant column have to be included?.. Or can the `tasks.json` be set up to so that the `round_id` is taken not from variable, but from filename?

**Reply by bsweger** [reply]
[This turned out longer than expected--happy to hop on a zoom if it's time to discuss synchronously!]

> Does the constant column have to be included?.. Or can the tasks.json be set up to so that the round_id is taken not from variable, but from filename?

After re-reading the [Hubverse docs on `tasks.json` and `round_id`](https://hubdocs.readthedocs.io/en/latest/quickstart-hub-admin/tasks-config.html#step-5-define-task-ids), I'm still not sure of the official Hubverse answer (hopefully @annakrystalli can weigh in).

The `model-output` data transform function currently assumes that `round_id` is part of the filename _and_ that it appears as a constant column in `model-output`. _I.e._, the resulting parquet file will have the constant value in two columns, each representing a different role it's playing (`epiweek`, and `round_id`). In other words:
* `round_id` is derived from the filename and added as a column
* the actual data represented by `round_id` (e.g., `epiweek`) is assumed to be in the original file, so there's no additional code to look for it elsewhere and add it.

I think it's important to include both columns in the transformed data (even if repetitive) to ensure ease of use for people not familiar with the Hubverse and its config files.  From my perspective, if space and not coding complexity is the only drawback, it would be more straightforward to include it as a column when porting over the old data. While we can (and should) update the transform function to recognize various ways that `task.json` can define a `round_id`, I'd prefer to get the simpler version of the transform up and running before attempting to codify various ways the combination of round_id/task.json/variable can work.

> Note that model-output folder has a subfolder for each model (the name of that folder is sometimes in team-model format (for example "NEU-GLEAM", but in many instances that type of information was not provided, so we have folder names that really just describe the model itself

Got it. Since you're doing the work to get this historic data into Hubverse format, I think it makes sense to follow the guidance which [specifies a `team-model` format for the folder names](https://hubdocs.readthedocs.io/en/latest/user-guide/model-output.html#directory-structure). If there's no team name available, what if we just pop the model name in that field instead? That would ensure we don't break any downstream Hubverse tooling that expects `team-model`. E.g., `ARETE-ARETE` (or even `unknown-ARETE` if we don't want to use model name as a team placeholder).

------

Not quite related to this convo, but maybe worth surfacing...

Practically speaking, I _think_ the Hubverse validation function is [getting round_id from the filename _and_ expecting it in YYYY-MM-DD format](https://github.com/Infectious-Disease-Modeling-Hubs/hubValidations/blob/main/R/parse_file_name.R#L27). It's not clear that model-output file that uses a differently formatted round_id would pass validations, even though it's permitted according to the spec. [Related discussion](https://github.com/Infectious-Disease-Modeling-Hubs/hubValidations/discussions/13).

If you're not running the converted data through the validation function, this point might be irrelevant, but raising it just in case.

**Reply by annakrystalli** [reply]
Hi @lmullany !


@bsweger is correct that validation is currently set up expecting dates as round IDs as all current hubs use that format.  However from the beginning there was discussion of supporting different ways of configuring and validating more diverse round IDs. This historical hub seems a great opportunity to push forward with that!

Currently there is an option to not include round ID details in the file by setting `round_id_from_variable` to `false`. This does mean that a separate round configuration will need to be included for each round id though. This would especially be important because the submission window would need to manually be specified with a start and end date for each round given that a window cannot be calculated from a reference date variable within the file. Unless, out of curiosity, is some sort of refence date information included in each file? I'm also curious, does the epi week indicate the week the predictions relate to?

**Reply by annakrystalli** [reply]
Hello @lmullany,

I've put together an issue on supporting non-date round IDs which comes with some implications with respect to how non-date hubs are configured. See details in the issue: https://github.com/Infectious-Disease-Modeling-Hubs/hubValidations/issues/81

My suggestion regarding the hub you are building would be to in fact not use epiweeks as round ids but convert them to an origin date. This would allow you to specify submission windows relative to those dates, e.g.:
```json
            "submissions_due": {
                "relative_to": "origin_date",
                "start": -6,
                "end": 2
            }
```

Currently, in your tasks.json,[ the submission windows have not been appropriately specified](https://github.com/lmullany/flusight1_hub/blob/7201e9a72e8e7906f5bc781a1fa849d1471a701e/hub-config/tasks.json#L131-L134), and if you continue with origin epiweek codes for round ids, you would need to create a[ separate config block for each round](https://github.com/Infectious-Disease-Modeling-Hubs/hubValidations/issues/81) and round_id and manually specify the start and end date of each round submission window, e.g.:
```json
 "submissions_due": {
                    "start": "2022-12-20",
                    "end": "2022-12-28"
                }
```

So in my view, it would be easier to convert the hub to what the flusight hub is currently doing, i.e. using dates as `round_id`s.

Does that make sense?

**Reply by annakrystalli** [reply]
Having said the above, that's only if you are planning to get the hub to a state where it would pass new submission validation tests (i.e. would be fully configured and functional to the point of being able to accept new submissions). If that is not an aim of such a historic hub (i.e. just being able to access data through `hubData::connect_hub()` is enough), I think it would be fine to ignore the above suggestion and leave the hub as is. The only thing that would need to be addressed is the lack of consistently named `model-output` subdirectories but @bsweger's suggestion should fix that.

**Reply by lmullany** [reply]
@annakrystalli  all very helpful - thanks!!  I tend to think that your final comment (re just getting it to a state where the other downstream tools can be used appropriately, `connect_hub()` etc, can be used) resonates the most with me

On the other hand your suggestion to convert to a date based round id, is also a good one.  Given that we can make a 1-1 conversion of the current round id (2015-42, for example) to a date  (say, the first or last day of that epiweek), we could easily implement your suggestion.   However, I'm not sure re the submission due date.. We can put in submission due dates that are global to a date-based round id (like your above suggestion, replicated below)

```yaml
            "submissions_due": {
                "relative_to": "origin_date",
                "start": -6,
                "end": 2
            }
```
but I'm not familiar with what the actual rules were at the time (and many submissions would not be valid if checked against any specific submissions_due configuration that we put here in the config json files)


agree that we should certainly address the issues re model-output names and model-metadata

**Reply by annakrystalli** [reply]
Regarding the submission due property, I don't think it's a problem that retrospectively some submissions may not have met the submission window criteria. Indeed, even in current hubs, a mantainer may allow a submission outside the window to be merged in if they like.

 Having said that, moving to a date `round_id` and setting up a relative submission would mean that, **in theory**, if you ever wished to open  the hub up for submissions, all you would need to do is add to the optional values of `origin_date` and submission windows for new round_ids would be validated successfully.  It would also serve as an example of an actually functional and correctly configured hub.

I tend to now think though that a more important reason to convert to dates might be downstream functionality like scoring or visualising hub data but I'll let folks that are more familiar with those aspects (e.g. @elray1 , @LucieContamin @nickreich etc) weigh in on that.

In any case, if you do decide to go ahead with it, I found this snippet useful when playing around with configs for this hub:
``` r
epiweek <- "2022-52"
lubridate::parse_date_time(paste(epiweek, 1, sep = "-"), 'Y-W-w')
#> [1] "2022-12-26 UTC"
```

<sup>Created on 2024-04-26 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Reply by LucieContamin** [reply]
I am not sure I totally understand the issue here, sorry. 
But, for SMH, we mainly use `origin_date` as `round_id` . However, we have some rounds where the `round_id` is not the `origin_date`, and is only use in the filename, to be able to tag which file correspond to which round. In this case, the format of `round_id` does not matter a lot.  We still use a YYYY-MM-DD format to follow the same "style" as the other round.
Does that answer your question? or help?

**Reply by nickreich** [reply]
My feeling is that the largest advantage to changing the epiweeks to dates in the "round id" column and filenames would be to just make things more consistent with what we now see as "best practices" for hubs. It will save downstream analysts time wrangling date conversions like the one Anna did above. That said, it is tempting to just kind of leave it as is for simplicity's sake since if I understand correctly, @annakrystalli is saying that it won't break anything to have the EW filenames.

**Reply by annakrystalli** [reply]
Right, it won't break anything.  It's just that [there is no date column whatsoever in the data](https://github.com/lmullany/flusight1_hub/blob/main/hub-config/tasks.json), just epiweek codes so I imagine, plotting step ahead for example would be a two step process, converting epiweeks to origin date and then origin dates to target dates. Not massive effort I guess,  just needs to be documented somewhere so downstream analysts know what to do.( @LucieContamin that's what my question was referring to)

**Reply by nickreich** [reply]
Yeah, so I guess I would vote for resolving the epiweeks to an origin_date column then, for simplicity. Maybe we keep an epiweek column, for posterity?

**Reply by annakrystalli** [reply]
> I am not sure I totally understand the issue here, sorry. But, for SMH, we mainly use `origin_date` as `round_id` . However, we have some rounds where the `round_id` is not the `origin_date`, and is only use in the filename, to be able to tag which file correspond to which round. In this case, the format of `round_id` does not matter a lot. We still use a YYYY-MM-DD format to follow the same "style" as the other round. Does that answer your question? or help? 

@LucieContamin  This comment is actually really useful as a response to https://github.com/Infectious-Disease-Modeling-Hubs/hubValidations/issues/81

Do you mind if we transfer it over so I can ask follow up qustions there?

**Reply by LucieContamin** [reply]
please feel free to transfer it

**Reply by LucieContamin** [reply]
> Yeah, so I guess I would vote for resolving the epiweeks to an origin_date column then, for simplicity. Maybe we keep an epiweek column, for posterity?

Sorry I misunderstood! I agree that it might be helpful to add a date column and maybe keep the epiweek column so you don't need to change the round_id. Will that work?

**Reply by lmullany** [reply]
Most of all of this above belongs here: https://github.com/lmullany/flusight1_hub/issues/10

I'm going to move forward with creating a new column `origin_date`, and have started a new issue [here ](https://github.com/lmullany/flusight1_hub/issues/19) to discuss the exact derivation of that origin_date

**Reply by nickreich** [reply]
Some quotes from the 2017-2018 FluSight guidance docs:
> Forecasts for the ILINet percentage for the four weeks following the forecast submission should be relative to the most recent week of ILINet data released. For example, ILINet data for week 43 will be posted on Friday, November 3 at 12:00PM Eastern Time. The four-week forecast submitted on Monday, November 6 should include predictions for ILINet values for weeks 44-47.  

and 

> the filename should be modified to the following standard naming convention: a forecast submission using week 43 surveillance data submitted by John Doe University on November 6, 2017, should be named “EW43-JDU-2017-11-06.csv” where EW43 is the latest week of ILINet data used in the forecast, JDU is the name of the team making the submission (e.g. John Doe University), and 2017-11-06 is the date of submission.

I just wanted to clarify/confirm that the submission date is going to be used as the "round_id", and that we discussed yesterday that we might standardize this on being something like the last day (Saturday) of the epiweek prior to the Monday submission deadline? Whereas the origin date will likely be a week or so behind because of the ILI delays. see [the issue](https://github.com/lmullany/flusight1_hub/issues/19) @lmullany mentioned above for more details

**Reply by lmullany** [reply]
Actually, we are going to use the Saturday of the EWXX that is the prefix of each submission file, regardless of the submission date embedded in the filename. 

So, this is basically in line with what you have here, but I want to be explicit. This is also option 2 in the three options I laid out in the issue you mention above:

https://github.com/lmullany/flusight1_hub/issues/19

**Comment by annakrystalli** [comment]
Comment Moved to thread

**Comment by lmullany** [comment]
> Regarding the submission due property, I don't think it's a problem that retrospectively some submissions may not have met the submission window criteria. Indeed, even in current hubs, a mantainer may allow a submission outside the window to be merged in if they like.
> 
> Having said that, moving to a date `round_id` and setting up a relative submission would mean that, **in theory**, if you ever wished to open the hub up for submissions, all you would need to do is add to the optional values of `origin_date` and submission windows for new round_ids would be validated successfully. It would also serve as an example of an actually functional and correctly configured hub.
> 
> I tend to now think though that a more important reason to convert to dates might be downstream functionality like scoring or visualising hub data but I'll let folks that are more familiar with those aspects (e.g. @elray1 , @LucieContamin @nickreich etc) weigh in on that.
> 
> In any case, if you do decide to go ahead with it, I found this snippet useful when playing around with configs for this hub:
> 
> ```r
> epiweek <- "2022-52"
> lubridate::parse_date_time(paste(epiweek, 1, sep = "-"), 'Y-W-w')
> #> [1] "2022-12-26 UTC"
> ```
> 
> Created on 2024-04-26 with [reprex v2.0.2](https://reprex.tidyverse.org)

fyi @annakrystalli , I'm not sure that snippet is correct for convert epi weeks to a date. I use the following for that. Perhaps there is a better built in function that I'm not aware of. 
```r
#' Get start date of (epi) given epiweek and epiyear
#' 
#' Function returns the date of the first day of the epiweek defined by
#' user provided epiweek (`ew`) and epiyear (`ey`)

date_from_epiweekyear <- function(ey,ew) {

  # internal function gets the start date of the first week of the 
  # give epi year
  f_epi_week_start <- function(y) {
    j4 = strptime(paste0(y, "-01-04"), "%Y-%m-%d")
    dplyr::if_else(j4$wday == 7, as.Date(j4), as.Date(j4) - j4$wday)
  }
  
  # get the start date of this epi year
  start_1 = f_epi_week_start(ey)
  
  # check that the epi week passed by user does not exceed the
  # maximum available for this epi year
  max_week = (f_epi_week_start(ey+1)-start_1)/7
  if(ew>max_week) {
    "The provided epiweek number is not valid for this year"
    break
  }
  # return the start date of this epiweek/epiyear
  return(as.Date(start_1+(ew-1)*7))
}
```

**Reply by annakrystalli** [reply]
Hi @lmullany ,could you be more specific on why it is incorrect? According to this [table](https://www.epochconverter.com/weeks/2022) it returns the correct first day of the 52nd week in 2022 (26 Dec). If you need a different day of the week just change the 1 (which indicates the first day of the week) to whatever is needed.

**Reply by lmullany** [reply]
It is the fact that epiweeks are not  like regular weeks.  Its not just a matter of changing the `w` from 1 to 0. Try it for 2019-52

**Reply by lmullany** [reply]
for example
```r
y = 2019;w=52
```
Using your approach and confirming we can recover using `epiyear` and `epiweek` functions
```r
d1 = parse_date_time(paste(y,"-",w,0,sep="-"), "Y-W-w")
c(epiyear(d1), epiweek(d1))  

[1] 2020    1
```
the other approach:
```r
d2 = date_from_epiweekyear(y,w)
c(epiyear(d2), epiweek(d2))  

[1] 2019   52
```

**Reply by annakrystalli** [reply]
Ah ok! I don't think I'd properly understood what epiweeks were. That makes sense.

**Reply by LucieContamin** [reply]
The package `MMWRweek` is also very good at dealing with epiweeks, if that help.

For example:
```r
> MMWRweek::MMWRweek2Date(2019, 52)
[1] "2019-12-22"
> MMWRweek::MMWRweek2Date(2019, 52, 7)
[1] "2019-12-28"
```

**Reply by lmullany** [reply]
epiweeks are kind of annonying invention.. based on MMWR...

**Reply by lmullany** [reply]
yeah, lubridate should really offer a reverse function, otherwise you have to go to a number of other packages.. I just wanted to a quick implemention without installing any other package

**Reply by nickreich** [reply]
Agree with wanting this to be simple and the desire to not rely on outside packages. The logic you use is at least a bit different from the ["standard" MMWR implementation](https://github.com/cran/MMWRweek/blob/master/R/MMWRweek2Date.R). Do they produce the same results?

**Reply by lmullany** [reply]
I’ll check to be sure. Maybe I moved too fast and made some error. If so, I’ll fix.

**Reply by lmullany** [reply]
yep, it looks okay to me. I don't see any errors. Here is a much faster, more concise, and vectorized version that uses the same logic
```r
date_from_epiweekyear <- function(ey,ew) {
  # internal function gets the start date of the first week of year
  f <- \(y) { j4=paste0(y, "-01-04"); as.Date(j4) - wday(j4)+1}
  # get the start date of this epi year and next epi year
  s1 = f(ey); s2=f(ey+1)
  # check that max week is not exceeded  
  data.table::fifelse(ew>(f(ey+1) - s1)/7, NA_Date_, s1+(ew-1)*7)
}
```

Here is a quick test over the last couple of decades:
```r
library(data.table); library(lubridate)

df = data.table(date = seq.Date(as.Date("2001-01-01"), as.Date("2024-12-31"),1))
df[, c("ey", "ew", "ws"):={ey = epiyear(date); ew=epiweek(date);ws = date_from_epiweekyear(ey,ew); .(ey, ew,ws)}]

# Now, confirm that the ws is equal to date-wday()+1  
df[, all(ws==date-wday(date)+1)]
[1] TRUE
```

and, checking against MMWR week:
```r
df[, mmwr:=MMWRweek::MMWRweek2Date(ey, ew)][, all(ws==mmwr)]
[1] TRUE
```

**Reply by nickreich** [reply]
awesome!

**Comment by lmullany** [comment]
Question: how does the group feel about changing model name to all lower case (or all upper case). Similarly for team name. 
Examples:
 - `KoT-GPthingy` would become `kot-gpthingy`
 - `Delphi-Archefilter` would become `delphi-archefilter`

Not a requirement, but would look a bit more standard, and would help a bit as we are renaming / merging different misspellings, etc.

**Reply by bsweger** [reply]
yes, please!

**Comment by annakrystalli** [comment]
Definitely in favour of all lower case personally!
